cpu/stm32f1/periph/pwm.c

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (C) 2015 Engineering-Spirit
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32f1
+ * @{
+ *
+ * @file
+ * @brief       Low-level PWM driver implementation
+ *
+ * @author      Nick van IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include "cpu.h"
+#include "periph/pwm.h"
+#include "periph_conf.h"
+
+/* ignore file in case no PWM devices are defined */
+#if PWM_0_EN || PWM_1_EN
+
+int pwm_init(pwm_t dev, pwm_mode_t mode, unsigned int frequency, unsigned int resolution)
+{
+    TIM_TypeDef *tim = NULL;
+    GPIO_TypeDef *port = NULL;
+    uint32_t pins[PWM_MAX_CHANNELS];
+    uint32_t pwm_clk = 0;
+    int channels = 0;
+
+    pwm_poweron(dev);
+
+    switch (dev) {
+#if PWM_0_EN
+        case PWM_0:
+            tim = PWM_0_DEV;
+            port = PWM_0_PORT;
+            pins[0] = PWM_0_PIN_CH0;
+#if (PWM_0_CHANNELS > 1)
+            pins[1] = PWM_0_PIN_CH1;
+#endif
+#if (PWM_0_CHANNELS > 2)
+            pins[2] = PWM_0_PIN_CH2;
+#endif
+#if (PWM_0_CHANNELS > 3)
+            pins[3] = PWM_0_PIN_CH3;
+#endif
+            channels = PWM_0_CHANNELS;
+            pwm_clk = PWM_0_CLK;
+            PWM_0_PORT_CLKEN();
+            break;
+#endif
+
+#if PWM_1_EN
+        case PWM_1:
+            tim = PWM_1_DEV;
+            port = PWM_1_PORT;
+            pins[0] = PWM_1_PIN_CH0;
+#if (PWM_1_CHANNELS > 1)
+            pins[1] = PWM_1_PIN_CH1;
+#endif
+#if (PWM_1_CHANNELS > 2)
+            pins[2] = PWM_1_PIN_CH2;
+#endif
+#if (PWM_1_CHANNELS > 3)
+            pins[3] = PWM_1_PIN_CH3;
+#endif
+            af = PWM_1_PIN_AF;
+            channels = PWM_1_CHANNELS;
+            pwm_clk = PWM_1_CLK;
+            PWM_1_PORT_CLKEN();
+            break;
+#endif
+    }
+
+    /* setup pins for high speed alternate function */
+    for (int i = 0; i < channels; i++) {
+        /* calculate where the MODE and CNF must be updated */
+        port->CR[pins[i] >= 8] &= ~(0xfl << (4 * (pins[i] - ((pins[i] >= 8) * 8))));
```

Comment:
The F4 has a separate MODE and AFR register for the F1 it is configured via CR1 and CR2